### PR TITLE
Format macro invocations

### PR DESCRIPTION
**Pull request by marcusklaas**
Addresses https://github.com/nrc/rustfmt/issues/35.

Unformatted code is running out of places to hide!

### CODE REVIEW

**Comment by cassiersg** [comment]
Beautiful !

I prefer to leve the review to @nrc since I don't know the internals of the parser.

File: src/comment.rs [hunk 1]
```diff
@@ -300,10 +300,10 @@ mod test {
                                    12));
 
         let input = "// comment";
-        let expected = "/* com\n                                                                      \
-                        * men\n                                                                      \
-                        * t */";
-        assert_eq!(expected, rewrite_comment(input, true, 9, 69));
+        let expt = "/* com\n                                                                      \
```
Comment (nrc) [inline src/comment.rs]: nit: this seems like a worse name

File: tests/target/macros.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    vec![a, b, c];
+
+    vec![AAAAAA, AAAAAA, AAAAAA, AAAAAA, AAAAAA, AAAAAA, AAAAAA, AAAAAA, AAAAAA, BBBBB, 5,
+         100 - 30, 1.33, b, b, b];
+
+    vec![a /* comment */];
+
+    foo(makro!(1, 3));
+
+    macrowithbraces! {dont,    format, me}
```
Comment (nrc) [inline tests/target/macros.rs]: We should probably format macros with braces like structs

File: tests/target/macros.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    baz!(1 + 2 + 3, quux.kaas());
+
+    quux!(AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA,
+          BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB);
+
+    kaas!(// comments
+          a, // post macro
+          b /* another */);
+
+    noexpr!( i am not an expression, OK? );
```
Comment (nrc) [inline tests/target/macros.rs]: why do we keep the leading and trailing spaces here?

File: src/macros.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+use syntax::parse::{ParseSess, tts_to_parser};
+
+use rewrite::RewriteContext;
+use expr::{rewrite_call, rewrite_array};
+use comment::FindUncommented;
+use utils::wrap_str;
+
+// FIXME: this is a pretty terrible hack. Any other solution would be preferred.
+struct ForceSend<T>(pub T);
+unsafe impl<T> Send for ForceSend<T> {}
```
Comment (nrc) [inline src/macros.rs]: why do we need this? What are we sending which doesn't impl Send?

File: src/macros.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    } else if tt_vec.is_empty() {
+        return if let MacroStyle::Parens = style {
+            Some(format!("{}()", macro_name))
+        } else {
+            Some(format!("{}[]", macro_name))
+        };
+    }
+
+    let wrapped_tt_vec = ForceSend((*tt_vec).clone());
+    let expr_vec_result = thread::catch_panic(move || {
```
Comment (nrc) [inline src/macros.rs]: COuld do with a comment explaining why we're using a thread here

File: src/macros.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+fn macro_style(mac: &ast::Mac, context: &RewriteContext) -> MacroStyle {
+    let snippet = context.snippet(mac.span);
+
+    match (snippet.find_uncommented("("), snippet.find_uncommented("[")) {
+        (Some(i), Some(j)) if i < j => MacroStyle::Parens,
+        (Some(_), None) => MacroStyle::Parens,
+        (None, None) => MacroStyle::Braces,
+        _ => MacroStyle::Brackets,
```
Comment (nrc) [inline src/macros.rs]: This looks like it would fail for `foo!{ () }`

**Comment by nrc** [comment]
looks good, a few comments in line

File: tests/target/macros.rs (hunk 3 above)
Comment (marcusklaas) [inline tests/target/macros.rs]: Since it's not list-like, we just leave it as is.

File: src/macros.rs (hunk 4 above)
Comment (marcusklaas) [inline src/macros.rs]: A `TokenTree` is not send, since it can contains `Rc<T>` amongst other non-`Send`s.

File: tests/target/macros.rs (hunk 2 above)
Comment (marcusklaas) [inline tests/target/macros.rs]: Mind if I leave this for an other PR?

File: src/comment.rs (hunk 1 above)
Comment (marcusklaas) [inline src/comment.rs]: It is. I cheated a bit to fit the line within the limit, but I now realize there's better ways of achieving that.

**Comment by marcusklaas** [comment]
Updated!

**Comment by nrc** [comment]
thanks for the changes!

**Pull request merged**